File: Haskell/Problem_023.hs
{-# LANGUAGE BangPatterns #-}
-- Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.
import           System.IO
import           Euler                          ( divisors )
import qualified Data.Set                      as Set -- for faster lookups


upperBoundSum2Abundants :: Integer
upperBoundSum2Abundants = 28123


abundants :: [Integer]
abundants = [ n | n <- [1 ..], (sum . divisors $ n) > n ]


abundantAddends :: Set.Set Integer
abundantAddends =
  Set.fromList $ takeWhile (< upperBoundSum2Abundants) abundants


sumNonSum2Abundants :: Integer
sumNonSum2Abundants = sum nonSum2Abundants
 where
  nonSum2Abundants =
    filter (not . isSumAbundants) [1 .. upperBoundSum2Abundants]
  isSumAbundants n =
    any (\ab -> (n - ab) `Set.member` abundantAddends) abundantAddends


main :: IO ()
main = do
  print $ sumNonSum2Abundants

File: Python 3/Problem_023.py
# Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.
from euler import divisors
from itertools import takewhile

UPPER_BOUND_SUM_TWO_ABUNDANTS = 28124


def abundants():
    n = 1
    while True:
        if sum(divisors(n)) > n:
            yield n
        n += 1


ABUNDANT_ADDENDS = set(
    takewhile(
        lambda ab: ab < UPPER_BOUND_SUM_TWO_ABUNDANTS,
        (abundant for abundant in abundants()),
    )
)


def not_sum_2_abundants():
    for n in range(1, UPPER_BOUND_SUM_TWO_ABUNDANTS):
        if not any(n - ab in ABUNDANT_ADDENDS for ab in ABUNDANT_ADDENDS):
            yield n


if __name__ == "__main__":
    print(sum(i for i in not_sum_2_abundants()))
